File: src/main/java/pl/swmud/ns/swaedit/gui/NewShopWidget.java
package pl.swmud.ns.swaedit.gui;

import java.math.BigInteger;

import pl.swmud.ns.swmud._1_0.area.Mobiles.Mobile;
import pl.swmud.ns.swmud._1_0.area.Repairs.Repair;
import pl.swmud.ns.swmud._1_0.area.Shops.Shop;

import com.trolltech.qt.core.Qt;
import com.trolltech.qt.gui.QWidget;

class NewShopWidget extends QWidget {
    
    enum Type {
        SHOP,
        REPAIR
    }

    private Ui_NewShopWidget ui = new Ui_NewShopWidget();
    private boolean found;
    Signal1<BigInteger> vnumChosen = new Signal1<BigInteger>();
    
    public NewShopWidget(Type type) {
        ui.setupUi(this);
        SWAEdit.setChildPosition(this);
        setAttribute(Qt.WidgetAttribute.WA_DeleteOnClose);
        setWindowModality(Qt.WindowModality.ApplicationModal);

        if (type == Type.REPAIR) {
            for (Mobile mob : SWAEdit.ref.area.getMobiles().getMobile()) {
                found = false;
                for (Repair repair : SWAEdit.ref.area.getRepairs().getRepair()) {
                    if (mob.getVnum().equals(repair.getKeeper())) {
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    ui.keeperBox.addItem(mob.getVnum()+" - "+mob.getName(), mob.getVnum());
                }
            }
        }
        else
        {
            for (Mobile mob : SWAEdit.ref.area.getMobiles().getMobile()) {
                found = false;
                for (Shop shop : SWAEdit.ref.area.getShops().getShop()) {
                    if (mob.getVnum().equals(shop.getKeeper())) {
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    ui.keeperBox.addItem(mob.getVnum()+" - "+mob.getName(), mob.getVnum());
                }
            }
        }
        
        ui.keeperBox.setCurrentIndex(0);
    }
    
    @SuppressWarnings("unused")
    private void on_cancelButton_clicked() {
        close();
    }

    @SuppressWarnings("unused")
    private void on_acceptButton_clicked() {
        hide();
        vnumChosen.emit((BigInteger)ui.keeperBox.itemData(ui.keeperBox.currentIndex()));
        close();
    }
}

File: src/gui/NewShopWidget.cpp
#include "gui/NewShopWidget.h"
#include "ui_NewShopWidget.h"

NewShopWidget::NewShopWidget(Area *area, bool isRepair, QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::NewShopWidget)
    , area_(area)
    , isRepair_(isRepair)
{
    ui->setupUi(this);
    setWindowFlag(Qt::Window);
    setAttribute(Qt::WA_DeleteOnClose);
    setWindowModality(Qt::ApplicationModal);

    if (isRepair_) {
        setWindowTitle("New Repair Shop");
    } else {
        setWindowTitle("New Shop");
    }

    // Fill keeper box with mobiles that don't already have a shop (or repair)
    ui->keeperBox->clear();
    for (const Mobile &mob : area_->mobiles) {
        bool alreadyHasShop = false;

        if (isRepair_) {
            for (const Repair &r : area_->repairs) {
                if (r.keeper == mob.vnum) {
                    alreadyHasShop = true;
                    break;
                }
            }
        } else {
            for (const Shop &s : area_->shops) {
                if (s.keeper == mob.vnum) {
                    alreadyHasShop = true;
                    break;
                }
            }
        }

        if (!alreadyHasShop) {
            ui->keeperBox->addItem(
                QString::number(mob.vnum) + " - " + mob.name,
                QVariant(mob.vnum));
        }
    }

    connect(ui->acceptButton, &QPushButton::clicked, this, &NewShopWidget::onAcceptClicked);
    connect(ui->cancelButton, &QPushButton::clicked, this, &NewShopWidget::onCancelClicked);
}

NewShopWidget::~NewShopWidget()
{
    delete ui;
}

void NewShopWidget::onAcceptClicked()
{
    if (ui->keeperBox->currentIndex() < 0) return;

    qint64 keeperVnum = ui->keeperBox->currentData().toLongLong();
    emit vnumChosen(keeperVnum);
    close();
}

void NewShopWidget::onCancelClicked()
{
    close();
}
